### PR TITLE
Box `ast::ItemKind`

### PR DESCRIPTION
**Pull request by jyn514**
This reduces the size of ast::Item from 296 to 96 bytes.

See https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/windows-rs.20perf/near/223974946 for context; the hope is this will save memory compiling windows-rs.

### CODE REVIEW

**Comment by ThePuzzlemaker** [comment]
The fact that this has not been done before genuinely scares me.

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -2593,6 +2593,10 @@ pub struct Item<K = ItemKind> {
     pub tokens: Option<LazyTokenStream>,
 }
 
+rustc_data_structures::static_assert_size!(Item<ItemKind>, 96);
```
Comment (cynecx) [inline compiler/rustc_ast/src/ast.rs:2596]: Does this account for pointer size differences across platforms?

**Comment by bugadani** [comment]
Well this has the exact opposite effect: increasing memory use. On the flip side, a small perf gain.

**Comment by bugadani** [comment]
The biggest `ItemKind` offending variants are:

* `Impl`: 200 bytes
* `Fn`: 176 bytes
* `Trait`: 136 bytes
* `TyAlias`: 120 bytes

**Comment by davidtwco** [comment]
Should this be closed in favour of #81405 or are there more things you'd like to try here @jyn514?

**Comment by jyn514** [comment]
I'm not planning to follow up on this.

**Pull request closed without merging**